pzsd_bot/__main__.py

```diff
@@ -39,10 +54,132 @@ async def on_message(message):
+                result = await conn.execute(
+                    select(pzsd_user).where(
+                        pzsd_user.c.discord_snowflake == recipient_id
+                    )
+                )
+            try:
+                recipient = result.one()
+            except NoResultFound:
+                logger.info(
+                    "%s tried to bestow points to '%s' but they weren't in the user table",
+                    bestower.name,
+                    recipient_name or recipient_id,
+                )
+                return
+
+        valid_transaction = True
+        if bestower.id == recipient.id:
+            valid_transaction = False
+            logger.info(
+                "%s attempted to give themselves %s points. Very naughty.",
+                bestower.name,
+                point_amount,
+            )
+        else:
+            logger.info(
+                "%s awarding %s point(s) to %s",
+                bestower.name,
+                point_amount,
+                recipient.name,
+            )
+
+        if valid_transaction:
+            async with engine.begin() as conn:
+                await conn.execute(
+                    insert(ledger).values(
+                        bestower=bestower.id,
+                        recipient=recipient.id,
+                        points=point_amount,
+                    )
+                )
+                logger.info("Added point transaction to ledger")
+
+        if valid_transaction:
+            title = "Point transaction"
+            color = 0xFFFFFF
+        else:
+            title = "Self point violation!"
+            color = 0xFF0000
+        embed = discord.Embed(
+            title=title,
+            description=f"[Jump to original message]({message.jump_url})",
+            colour=color,
+            timestamp=datetime.now(),
+        )
+        embed.add_field(name="Bestower", value=bestower.name, inline=True)
+        embed.add_field(name="Recipient", value=recipient.name, inline=True)
+        embed.add_field(name="Point amount", value=str(point_amount), inline=True)
+        message_content = message.content
+        if len(message_content) > 80:
+            message_content = message_content[:80] + "..."
+        embed.add_field(name="Content of message:", value=message_content, inline=False)
+
+        points_log_channel = bot.get_channel(int(os.environ["POINTS_LOG_CHANNEL"]))
+        await points_log_channel.send(embed=embed)
+
 
 @bot.slash_command(guild_ids=[GUILD_ID])
-async def test(ctx):
-    await ctx.respond("This was a test.")
+async def leaderboard(ctx):
+    points = {}
+
+    async with engine.connect() as conn:
+        result = await conn.execute(select(pzsd_user))
+        for row in result.fetchall():
+            points[row.id] = {"name": row.name, "point_total": 0}
+
+        result = await conn.execute(select(ledger))
+        for row in result:
+            points[row.recipient]["point_total"] += row.points
```

Comment:
Use the in-database sum function

Reply:
How? I was thinking about this, and would that not require 5 separate queries (one for each user)? I couldn't think of an efficient query to individually sum up each person's points, is it possible to do that in one single query?

Reply:
```sql
select
  pzsd_user.name,
  sum(ledger.points)
from
  ledger
  left outer join pzsd_user on pzsd_user.id = ledger.recipient;
group by
  pzsd_user.id;
```

Reply:
clever, i'll definitely change it to this